src/snowflake/snowpark/mock/_stored_procedure.py

```diff
@@ -0,0 +1,238 @@
+    StoredProcedure,
+    StoredProcedureRegistration,
+)
+from snowflake.snowpark.types import (
+    ArrayType,
+    DataType,
+    MapType,
+    StructType,
+    _FractionalType,
+    _IntegralType,
+)
+
+from ._telemetry import LocalTestOOBTelemetryService
+
+if sys.version_info <= (3, 9):
+    from typing import Iterable
+else:
+    from collections.abc import Iterable
+
+
+def sproc_types_are_compatible(x, y):
+    if (
+        isinstance(x, type(y))
+        or isinstance(x, _IntegralType)
+        and isinstance(y, _IntegralType)
+        or isinstance(x, _FractionalType)
+        and isinstance(y, _FractionalType)
+    ):
+        return True
+    return False
+
+
+class MockStoredProcedure(StoredProcedure):
+    def __call__(
+        self,
+        *args: Any,
+        session: Optional["snowflake.snowpark.session.Session"] = None,
+        statement_params: Optional[Dict[str, str]] = None,
+    ) -> Any:
+        args, session = self._validate_call(args, session)
+
+        # Unpack columns if passed
+        parsed_args = []
+        for arg, expected_type in zip(args, self._input_types):
+            if isinstance(arg, Column):
+                expr = arg._expression
+
+                # If expression does not define its datatype we cannot verify it's compatibale.
+                # This is potentially unsafe.
+                if expr.datatype and not sproc_types_are_compatible(
+                    expr.datatype, expected_type
+                ):
+                    raise ValueError(
+                        f"Unexpected type {expr.datatype} for sproc argument of type {expected_type}"
+                    )
+
+                # Expression may be a nested expression. Expression should not need any input data
+                # and should only return one value so that it can be passed as a literal value.
+                # We pass in a single None value so that the expression evaluator has some data to
+                # pass to the expressions.
+                resolved_expr = calculate_expression(
+                    expr,
+                    ColumnEmulator(data=[None]),
+                    session._analyzer,
+                    {},
+                )
+
+                # If the length of the resolved expression is not a single value we cannot pass it as a literal.
+                if len(resolved_expr) != 1:
+                    raise ValueError(
+                        "[Local Testing] Unexpected argument type {expr.__class__.__name__} for call to sproc"
+                    )
+                parsed_args.append(resolved_expr[0])
+            else:
+                parsed_args.append(arg)
+
+        result = self.func(session, *parsed_args)
+
+        # Semi-structured types are serialized in json
+        if isinstance(
+            self._return_type,
+            (
+                ArrayType,
+                MapType,
+                StructType,
+            ),
+        ) and not isinstance(result, DataFrame):
```

Comment:
We have no support for table sprocs yet, why do we need `not isinstance(result, DataFrame)`?

Reply:
I've seen examples in the tests where a sproc returns a Dataframe, but it's signature says it returns a Structtype. If it's truly a struct type it needs to be serialized, but if it's just a dataframe it does not.

Reply:
I see. Could you please point me to the test?

Reply:
See [here](https://github.com/snowflakedb/snowpark-python/blob/e7421b44623e4388fe4f31596ca65bd45abf8109/tests/integ/test_stored_procedure.py).